blockstore/cassbs/cassandra_ds.go

```diff
@@ -0,0 +1,230 @@
+	var value []byte
+	err := cds.session.Query("SELECT value FROM lotus.chain WHERE key = ?", toCasKey(key)).WithContext(ctx).Scan(&value)
+	if err != nil {
+		if err == gocql.ErrNotFound {
+			return nil, datastore.ErrNotFound
+		}
+		return nil, err
+	}
+	return value, nil
+}
+
+func (cds *CassandraDatastore) Has(ctx context.Context, key datastore.Key) (bool, error) {
+	var count int
+	err := cds.session.Query("SELECT COUNT(*) FROM lotus.chain WHERE key = ?", toCasKey(key)).WithContext(ctx).Scan(&count)
+	if err != nil {
+		return false, err
+	}
+	return count > 0, nil
+}
+
+func (cds *CassandraDatastore) GetSize(ctx context.Context, key datastore.Key) (int, error) {
+	value, err := cds.Get(ctx, key) // todo this is not great, but getting blob len is not that easy. Hopefully we don't call this much
+	if err != nil {
+		return -1, err
+	}
+	return len(value), nil
+}
+
+func (cds *CassandraDatastore) Query(ctx context.Context, q query.Query) (query.Results, error) {
+	// Basic implementation assumes all filters, orders and limits are applied client-side
+	// todo do more fancy things if needed
+	iter := cds.session.Query("SELECT key, value FROM lotus.chain").WithContext(ctx).Iter()
+
+	var (
+		k       string
+		v       []byte
+		entries []query.Entry
+	)
+
+	for iter.Scan(&k, &v) {
+		vs := string(v) // copy
+		v := []byte(vs)
+
+		k, err := base64.StdEncoding.DecodeString(k)
+		if err != nil {
+			return nil, xerrors.Errorf("decode key: %w", err)
+		}
+
+		entry := query.Entry{Key: string(k), Value: v}
+		entries = append(entries, entry)
+	}
+	sort.Slice(entries, func(i, j int) bool { // todo eww
+		return entries[i].Key < entries[j].Key
+	})
+
+	if err := iter.Close(); err != nil {
+		return nil, err
+	}
+
+	// Apply filters, orders and limits client-side
+	qr := query.ResultsWithEntries(q, entries)
+	for _, filter := range q.Filters {
+		qr = query.NaiveFilter(qr, filter)
+	}
+	qr = query.NaiveOrder(qr, q.Orders...)
+	qr = query.NaiveLimit(qr, q.Limit)
+	qr = query.NaiveOffset(qr, q.Offset)
+
+	return qr, nil
+}
+
+var _ datastore.Read = (*CassandraDatastore)(nil)
+
+func (cds *CassandraDatastore) Sync(ctx context.Context, prefix datastore.Key) error {
+	return nil
+}
+
+func (cds *CassandraDatastore) Close() error {
+	cds.session.Close()
+	return nil
+}
+
+var _ datastore.Datastore = (*CassandraDatastore)(nil)
+
+type cassandraBatch struct {
+	session *gocql.Session
+	batch   *gocql.Batch
+}
+
+func (c *cassandraBatch) Put(ctx context.Context, key datastore.Key, value []byte) error {
+	statement := "UPDATE lotus.chain SET value = ? WHERE key = ?"
+	c.batch.Query(statement, value, toCasKey(key))
+	return nil
+}
+
+func (c *cassandraBatch) Delete(ctx context.Context, key datastore.Key) error {
+	statement := "DELETE FROM lotus.chain WHERE key = ?"
+	c.batch.Query(statement, toCasKey(key))
+	return nil
+}
+
+func (c *cassandraBatch) Commit(ctx context.Context) error {
+	var err error
+	for i := 0; i < 30; i++ {
```

Comment:
what's the meaning of looping 30 times here?

Reply:
This just retries very hard